### ISSUE TITLE
'Singles' puzzle: Whole grid is coloured red when disconnected, in latest beta release

### ISSUE THREAD

**Issue by SteffenBauer** [chrisboyle/sgtpuzzles#622]
**Describe the bug**
In the *singles* puzzle, as of the rules the grid of white spaces must be connected.

In the previous app releases, an disconnected grid was indicated by coloring in red only the areas that are disconnected from the main area.

But in the [latest beta release](https://github.com/chrisboyle/sgtpuzzles/releases/tag/2023-10-15-1359-2d9e414e), all white cells are now coloured red when the grid is disconnected.

**To Reproduce**
1. Install release [2023-10-15-1359-2d9e414e](https://github.com/chrisboyle/sgtpuzzles/releases/tag/2023-10-15-1359-2d9e414e)
2. Start any new *singles* puzzle
3. Add black spaces until the grid gets disconnected
4. Observed effect: All white cells become coloured red.

**Expected behavior**
Only the smallest separated grid areas should become colored red to indicate the disconnection error.

**Screenshots**

**Singles puzzle in latest beta release with disconnected grid** (incorrect behaviour):

![singles_wrong](https://github.com/chrisboyle/sgtpuzzles/assets/5973070/fbbf6665-4b24-49df-a831-31fabb584081)

**Singles puzzle in earlier releases with same saved game** (expected behaviour):

![singles_correct](https://github.com/chrisboyle/sgtpuzzles/assets/5973070/78218c25-b8a5-4a7a-b4c2-c0c0f7d764f7)

**Additional context**
Possible cause for the bug, and possible solution:

The bug might be caused by line 551 in singles.c. The code there makes the effort to identify the largest connected area as the 'canonical' one, so I think that colouring the whole grid in red is really not the expected behaviour after all.

I suspect it erroneously saves the grid number `i` as canonical number and not `dsf_canonify(dsf, i)`. It might be that this was a bug all the years in the code, which somehow became only visible when the DSF module was heavily refactored in upstream since the last release.

The same bug appears in my local build of the PC puzzle collection on Linux. When I change the line corresponding to line 551 in the android `singles.c` codefile from
```
if (largest < size) {
    largest = size;
    canonical = i;
}
```
to
```
if (largest < size) {
    largest = size;
   canonical = dsf_canonify(dsf, i);
}
```
then the error colouring is back to the expected behaviour.

**Comment by x-sheep** [comment]
This is definitely an upstream issue.

**Comment by chrisboyle** [comment]
@SteffenBauer Thanks for the detailed bug report, I agree with your analysis and have sent Simon a link to this.

**Comment by x-sheep** [comment]
The fix is now available on the upstream.

**Comment by chrisboyle** [comment]
Fixed in commit 35f796542ee5d7e775643f526973aa75f483bf20, now merged
